### PR TITLE
Handle wrong message type from HMI was added

### PR DESCRIPTION
**Pull request by ZhdanovP**
Fixes [#1865](https://github.com/smartdevicelink/sdl_core/issues/1865)

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
- Provide unit tests
- Provide ATF scripts

### Summary
* In case: SDL cuts off fake parameters from response (request) that SDL
should use internally and this response (request) is invalid by any
reason, SDL log the corresponding error and ignore this response
(request)

* In case: HMI sends request (response, notification) with fake
parameters, SDL cut off fake parameters.

* In case: SDL cuts off fake parameters AND RPC becomes is invalid,
SDL send GENERIC_ERROR.

* In case: HMI sends request without at least one mandatory param,
SDL log corresponding error internally and respond with 'INVALID_DATA'
to HMI

* In case: HMI sends request to SDL and this request has at least one
empty String param, SDL log corresponding error internally and respond
with 'INVALID_DATA' to HMI

Invalid response means the response contains:
- params out of bounds
- mandatory params are missing
- params of wrong type
- invalid json

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h [hunk 1]
```diff
@@ -145,6 +145,13 @@ class RPCHandlerImpl : public RPCHandler,
 #endif  // TELEMETRY_MONITOR
 
  private:
+  /**
+   * @brief Checks if message has to be sent to mobile or not
+   * @param output - message to check
+   * @returns true if message type is response otherwise false
+   */
+  bool HandleWrongMessageType(smart_objects::SmartObject& output) const;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @ZhdanovP I believe this method could be const

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h [hunk 2]
```diff
@@ -145,6 +145,13 @@ class RPCHandlerImpl : public RPCHandler,
 #endif  // TELEMETRY_MONITOR
 
  private:
+  /**
+   * @brief Checks if message has to be sent to mobile or not
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @ZhdanovP Unclear description.

File: src/components/application_manager/src/rpc_handler_impl.cc [hunk 3]
```diff
@@ -277,7 +277,8 @@ bool RPCHandlerImpl::ConvertMessageToSO(
                     "Convertion result: "
                         << result << " function id "
                         << output[jhs::S_PARAMS][jhs::S_FUNCTION_ID].asInt());
-      if (!hmi_so_factory().attachSchema(output, false)) {
+      if (!hmi_so_factory().attachSchema(output,
+                                         /*remove_fake_parameters=*/true)) {
```
Comment (AByzhynar) [inline src/components/application_manager/src/rpc_handler_impl.cc]: @ZhdanovP Remove this ugly comment.

File: src/components/application_manager/src/rpc_handler_impl.cc [hunk 4]
```diff
@@ -345,6 +342,47 @@ bool RPCHandlerImpl::ConvertMessageToSO(
...
+      output[strings::params][strings::message_type] =
+          MessageType::kErrorResponse;
+      output[strings::params][strings::error_msg] = "";
+      return false;
+    }
+    case application_manager::MessageType::kResponse: {
+      LOG4CXX_ERROR(logger_, "Received invalid data on HMI response");
+      break;
+    }
+    case application_manager::MessageType::kUnknownType: {
```
Comment (AByzhynar) [inline src/components/application_manager/src/rpc_handler_impl.cc]: @ZhdanovP May be it is worth to add `default` section for this case?

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h (hunk 1 above)
Comment (ZhdanovP) [reply src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @AKalinich-Luxoft This method already const

File: src/components/application_manager/src/rpc_handler_impl.cc (hunk 4 above)
Comment (ZhdanovP) [reply src/components/application_manager/src/rpc_handler_impl.cc]: @AByzhynar  Yes, good point

File: src/components/application_manager/src/rpc_handler_impl.cc (hunk 3 above)
Comment (ZhdanovP) [reply src/components/application_manager/src/rpc_handler_impl.cc]: @AByzhynar fixed f3ef0f9285e8dfa8e4ae1607038b9b73f5b4f1fc

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h (hunk 2 above)
Comment (ZhdanovP) [reply src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @AByzhynar fixed f3ef0f9285e8dfa8e4ae1607038b9b73f5b4f1fc

File: src/components/application_manager/src/rpc_handler_impl.cc (hunk 4 above)
Comment (AByzhynar) [reply src/components/application_manager/src/rpc_handler_impl.cc]: @ZhdanovP I did mean instead of this `case`

**Comment by mkorniichuk** [comment]
rebuild required

**Comment by mkorniichuk** [comment]
Rebuild required

**Comment by LitvinenkoIra** [comment]
Rebased to the latest develop (e137556)

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h (hunk 1 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @ZhdanovP please add description for `report`

**Comment by IGapchuk** [comment]
Rebased to the latest develop (49c29364460faa527bc758d0fd6b108463f8d82d)

**Comment by LitvinenkoIra** [comment]
@theresalech This PR is ready for Livio review

**Pull request merged**